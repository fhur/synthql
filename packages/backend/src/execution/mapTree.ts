interface Node {
    children: Node[];
}

interface Tree<TNode extends Node> {
    root: TNode;
}

export async function* mapTree<TInputNode extends Node, TOutputNode extends Node>(
    inputTree: Tree<TInputNode>,
    mapper: (inputNode: TInputNode, parentNode: TOutputNode | undefined) => Promise<TOutputNode>
): AsyncGenerator<Tree<TOutputNode>> {
    async function mapWithoutChildren(node: TInputNode, parentNode: TOutputNode | undefined): Promise<TOutputNode> {
        const result = await mapper(node, parentNode);
        result.children = [];
        return result;
    }

    // Initialize the queue with the root node and its mapped version.
    const queue: Array<{ originalNode: TInputNode; mappedNode: TOutputNode; parentNode?: TOutputNode }> = [];
    const mappedRoot = await mapWithoutChildren(inputTree.root, undefined);
    const outputTree: Tree<TOutputNode> = { root: mappedRoot };

    yield outputTree;

    queue.push({ originalNode: inputTree.root, mappedNode: mappedRoot });

    while (queue.length > 0) {

        const { originalNode, mappedNode, parentNode } = queue.shift()!;

        // Process all children of the current node in parallel.
        const mappedChildren = await Promise.all(
            originalNode.children.map((child) => {
                return mapWithoutChildren(child as TInputNode, parentNode)
            })
        );

        // After mapping, update the children of the current mapped node.
        mappedNode.children = mappedChildren;

        // Enqueue the children for their own processing, along with their mapped versions.
        originalNode.children.forEach((child, index) => {
            queue.push({
                originalNode: child as TInputNode,
                mappedNode: mappedChildren[index],
                parentNode: mappedNode
            });
        });

        // Yield the tree after each node is processed.
        yield outputTree;
    }

    // The function implicitly returns when the generator completes.
}

