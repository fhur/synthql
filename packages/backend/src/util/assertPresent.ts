
export function assertPresent<K, T extends K | null | undefined>(obj: T): asserts obj is NonNullable<T> {
    if (obj === null || obj === undefined) {
        throw new Error(`Expected value to be present but got: ${JSON.stringify(obj)}`);
    }
}
