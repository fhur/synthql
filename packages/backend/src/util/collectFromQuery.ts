import { AnyQuery } from "../types";

/**
 * Recursively maps a query and its children using the provided mapper function. 
 */
export function collectFromQuery<T>(query: AnyQuery, collect: (query: AnyQuery) => T[]): T[] {
    const collection: T[] = []
    const queue: AnyQuery[] = [query]

    while (queue.length > 0) {
        const current = queue.pop()!;
        collection.push(...collect(current))
        queue.push(...Object.values(current.include ?? {}));
    }
    return collection;
}