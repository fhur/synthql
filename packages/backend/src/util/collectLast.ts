import { collectAsync } from "./collectAsync";

export async function collectLast<T>(gen: AsyncGenerator<T>): Promise<T> {
    const list = await collectAsync(gen);
    return list[list.length - 1];
}
