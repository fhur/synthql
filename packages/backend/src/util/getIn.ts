import { Path } from "../execution/types";
import { assertArrayAtPath } from "./assertArrayAtPath";
import { assertObject } from "./assertObject";

export function getIn(tree: unknown, path: Path): unknown[] {
    if (path.length === 0) {
        return [tree];
    }

    const result: Array<unknown> = [];

    const queue: Array<{
        parent: unknown,
        sliceIndex: number,
        resolvedPathSegment: string | number
    }> = resolvePathSegment(path.slice(0, 1), tree).map(pathSegment => {
        return {
            parent: tree,
            sliceIndex: 1,
            resolvedPathSegment: pathSegment
        }
    })

    while (queue.length > 0) {
        const head = queue.shift()!;
        const { parent, sliceIndex, resolvedPathSegment } = head;
        const child = getInPathSegment(parent, path.slice(0, sliceIndex), resolvedPathSegment)


        const isLastIndex = sliceIndex === path.length;

        if (child === undefined) {
            continue;
        }
        if (isLastIndex) {
            result.push(child);
        }
        else {
            const nextSliceIndex = sliceIndex + 1;
            const nextSlice = path.slice(0, nextSliceIndex);
            resolvePathSegment(nextSlice, child).map(nextPathSegment => {
                return {
                    parent: child,
                    sliceIndex: nextSliceIndex,
                    resolvedPathSegment: nextPathSegment
                }
            }).forEach(next => queue.push(next));
        }
    }

    return result
}

function resolvePathSegment(path: Path, tree: unknown): Array<string | number> {
    const segment = path[path.length - 1];
    if (typeof segment === 'number') {
        return [segment];
    }
    if (typeof segment === 'string') {
        return [segment];
    }
    if (typeof segment === "object" && segment.type === 'anyIndex') {
        assertArrayAtPath(tree, path);
        return tree.map((_, i) => i);
    }
    throw new Error(`Unknown path segment: ${segment} at path ${path}`)
}

function getInPathSegment<T>(tree: T, path: Path, pathSegment: string | number): unknown | undefined {
    if (path.length === 0) {
        return tree;
    }

    if (tree === undefined) {
        return undefined;
    }

    if (typeof pathSegment === 'number') {
        assertArrayAtPath(tree, path);
        return tree[pathSegment];
    }
    if (typeof pathSegment === 'string') {
        assertObject(tree, path);
        return tree[pathSegment];
    }
    throw new Error(`Unknown path segment: ${pathSegment} at path ${path}`)
}