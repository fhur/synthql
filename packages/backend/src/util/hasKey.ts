import { isObj } from "./isObj";



export function hasKey<T extends string>(obj: unknown, key: T): obj is {
    [k in T]: unknown;
} {
    if (!isObj(obj)) {
        return false;
    }
    return key in obj;
}
