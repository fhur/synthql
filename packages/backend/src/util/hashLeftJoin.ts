/**
 * Performs a LEFT JOIN on two arrays using a hash table.
 */
function hashLeftJoin<TLeft, TRight>(
    left: TLeft[],
    right: TRight[],
    leftKey: (l: TLeft) => string,
    rightKey: (r: TRight) => string,
): Array<[TLeft, TRight | undefined]> {
    const rightHash = new Map<string, TRight>();
    for (const r of right) {
        rightHash.set(rightKey(r), r);
    }
    const result: Array<[TLeft, TRight | undefined]> = [];
    for (const l of left) {
        const r = rightHash.get(leftKey(l));
        result.push([l, r]);
    }
    return result;
}
