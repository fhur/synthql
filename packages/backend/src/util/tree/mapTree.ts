export interface Node {
    children: Node[];
}

interface Tree<TNode extends Node> {
    root: TNode;
}

export async function* mapTree<
    TInputNode extends Node,
    TOutputNode extends Node,
>(
    inputTree: Tree<TInputNode>,
    mapper: (
        inputNode: TInputNode,
        parentNode: TOutputNode | undefined,
    ) => Promise<TOutputNode>,
): AsyncGenerator<Tree<TOutputNode>> {
    async function mapWithoutChildren(
        node: TInputNode,
        parentNode: TOutputNode | undefined,
    ): Promise<TOutputNode> {
        const result = await mapper(node, parentNode);

        // We need to clear the children because `yield` the tree from time to time,
        // and we don't want to yield the tree with children that are not processed yet.
        result.children = [];
        return result;
    }

    // Initialize an empty queue.
    const queue: Array<{
        originalNode: TInputNode;
        mappedNode: TOutputNode;
    }> = [];
    const mappedRoot = await mapWithoutChildren(inputTree.root, undefined);
    const outputTree: Tree<TOutputNode> = { root: mappedRoot };

    // Yield the tree after the root node is processed.
    yield outputTree;

    // Enqueue the root node for processing.
    // No parent node at the root level.
    queue.push({ originalNode: inputTree.root, mappedNode: mappedRoot });

    while (queue.length > 0) {
        const { originalNode, mappedNode } = queue.shift()!;

        // Process all children concurrently.
        const newChildren = await Promise.all(
            originalNode.children.map(async (child) => {
                const originalNode = child as TInputNode;
                return {
                    originalNode,
                    mappedNode: await mapWithoutChildren(
                        originalNode,
                        mappedNode,
                    ),
                };
            }),
        );

        // After mapping, update the children of the current mapped node.
        mappedNode.children = newChildren.map((child) => child.mappedNode);

        // Yield the tree after each node is processed.
        yield outputTree;

        // Enqueue all children for processing.
        queue.push(...newChildren);
    }

    // The function implicitly returns when the generator completes.
}
