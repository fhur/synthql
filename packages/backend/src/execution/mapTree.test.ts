import { describe, expect, test } from "vitest";
import { mapTree } from "./mapTree";
import { collectAsync } from "../util/collectAsync";

describe('mapTree', () => {
    test('identity', async () => {
        const inputNode = {
            value: '1',
            children: [
                {
                    value: '2',
                    children: [
                        {
                            value: '4',
                            children: []
                        }
                    ]
                },
                {
                    value: '3',
                    children: []
                }
            ]
        }
        const tree = {
            root: inputNode
        }

        function sleep(ms: number) {
            return new Promise(resolve => setTimeout(resolve, ms));
        }

        const g = mapTree(tree, async node => {
            await sleep(1)

            return {
                value: `mapped: ${node.value}`,
                children: node.children
            }
        })
        for await (const t of g) {
            expect(t.root).toMatchSnapshot()
        }
    })
})

