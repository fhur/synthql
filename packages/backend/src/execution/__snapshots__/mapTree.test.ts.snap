// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mapTree > identity 1`] = `
{
  "children": [],
  "value": "mapped: 1",
}
`;

exports[`mapTree > identity 2`] = `
{
  "children": [
    {
      "children": [],
      "value": "mapped: 2",
    },
    {
      "children": [],
      "value": "mapped: 3",
    },
  ],
  "value": "mapped: 1",
}
`;

exports[`mapTree > identity 3`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "value": "mapped: 4",
        },
      ],
      "value": "mapped: 2",
    },
    {
      "children": [],
      "value": "mapped: 3",
    },
  ],
  "value": "mapped: 1",
}
`;

exports[`mapTree > identity 4`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "value": "mapped: 4",
        },
      ],
      "value": "mapped: 2",
    },
    {
      "children": [],
      "value": "mapped: 3",
    },
  ],
  "value": "mapped: 1",
}
`;

exports[`mapTree > identity 5`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "value": "mapped: 4",
        },
      ],
      "value": "mapped: 2",
    },
    {
      "children": [],
      "value": "mapped: 3",
    },
  ],
  "value": "mapped: 1",
}
`;
